src/runtime/node/crypto/$cloudflare.ts

```diff
@@ -0,0 +1,173 @@
+import type nodeCrypto from "node:crypto";
+import unenvCrypto from "./index";
+
+// @ts-ignore typings are not up to date, but this API exists, see: https://github.com/cloudflare/workerd/pull/2147
+const workerdCrypto = process.getBuiltinModule("node:crypto");
+
+// TODO: Ideally this list is not hardcoded but instead is generated when the preset is being generated in the `env()` call
+//       This generation should use information from https://github.com/cloudflare/workerd/issues/2097
+export const {
+  DiffieHellman,
+  DiffieHellmanGroup,
+  Hash,
+  Hmac,
+  KeyObject,
+  checkPrime,
+  checkPrimeSync,
+  createDiffieHellman,
+  createDiffieHellmanGroup,
+  createHash,
+  createHmac,
+  createPrivateKey,
+  createPublicKey,
+  createSecretKey,
+  generateKey,
+  generateKeyPair,
+  generateKeyPairSync,
+  generateKeySync,
+  generatePrime,
+  generatePrimeSync,
+  getCiphers,
+  getCurves,
+  getDiffieHellman,
+  getFips,
+  getHashes,
+  getRandomValues,
+  hkdf,
+  hkdfSync,
+  pbkdf2,
+  pbkdf2Sync,
+  randomBytes,
+  randomFill,
+  randomFillSync,
+  randomInt,
+  randomUUID,
+  secureHeapUsed,
+  setEngine,
+  setFips,
+  subtle,
+  timingSafeEqual,
+} = workerdCrypto;
+
+export const {
+  Certificate,
+  Cipher,
+  // @ts-ignore undocumented public API
+  Cipheriv,
+  Decipher,
+  // @ts-ignore undocumented public API
+  Decipheriv,
+  ECDH,
+  Sign,
+  Verify,
+  X509Certificate,
+  constants,
+  createCipheriv,
+  createDecipheriv,
+  createECDH,
+  createSign,
+  createVerify,
+  diffieHellman,
+  getCipherInfo,
+  hash,
+  privateDecrypt,
+  privateEncrypt,
+  publicDecrypt,
+  publicEncrypt,
+  scrypt,
+  scryptSync,
+  sign,
+  verify,
+} = unenvCrypto;
```

Comment:
Would it make sense that we use `export { ... } from './index'` to improve the tree-shaking possibility (in case of ESM imports)?

Reply:
I believe that this will tree-shake just fine because everything is statically analyzable, but I'll verify.

Reply:
I think bundler would be confused because we first import as namespace then use (runtime) restructure. At least it cannot be a safe treeshake for it because namespace import can have side-effects as well as restructure

Reply:
there is indeed a problem with tree-shaking. take a look at this...

Destructured import usage:
- [current](https://rollupjs.org/repl/?version=4.18.0&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMiUyRiUyRiUyMFRSRUUtU0hBS0lORyUyMC0lMjBkZXN0cnVjdHVyZWQlMjBpbXBvcnRzJTIwLSUyMGJyb2tlbiU1Q25pbXBvcnQlMjAlN0IlMjBCdWZmZXIlMkMlMjBpc0FzY2lpJTIwJTdEJTIwZnJvbSUyMCcuJTJGYnVmZmVyLWNsb3VkZmxhcmUtcG9seWZpbGwuanMnJTNCJTVDbiU1Q24lMkYlMkYlMjBzaG91bGQlMjByZXRhaW4lMjBvbmx5JTIwJTYwaXNBc2NpaSU2MCUyMGZyb20lMjB1bmVudi1ub2RlLWJ1ZmZlci1wb2x5ZmlsbC5qcyU1Q24lMkYlMkYlMjBidXQlMjByZXRhaW5zJTIwYWxsJTIwZHVlJTIwdG8lMjAlNjBpbXBvcnQlMjB1bmVudkJ1ZmZlciUyMGZyb20lMjAnLiUyRnVuZW52LW5vZGUtYnVmZmVyLXBvbHlmaWxsLmpzJyUzQiU2MCU1Q25jb25zb2xlLmxvZyhCdWZmZXIlMkMlMjBpc0FzY2lpKSUzQiUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTJDJTIybmFtZSUyMiUzQSUyMm1haW4uanMlMjIlN0QlMkMlN0IlMjJjb2RlJTIyJTNBJTIyaW1wb3J0JTIwdW5lbnZCdWZmZXIlMjBmcm9tJTIwJy4lMkZ1bmVudi1ub2RlLWJ1ZmZlci1wb2x5ZmlsbC5qcyclM0IlNUNuJTVDbmNvbnN0JTIwd29ya2VyZEJ1ZmZlciUyMCUzRCUyMHByb2Nlc3MuZ2V0QnVpbHRpbk1vZHVsZSglNUMlMjJub2RlJTNBYnVmZmVyJTVDJTIyKSUzQiU1Q24lNUNuZXhwb3J0JTIwY29uc3QlMjAlN0IlMjBCdWZmZXIlMkMlMjBTbG93QnVmZmVyJTdEJTIwJTNEJTVDbiUyMCUyMHdvcmtlcmRCdWZmZXIlM0IlNUNuJTVDbmV4cG9ydCUyMGNvbnN0JTIwJTdCJTIwaXNBc2NpaSUyQyUyMGlzVXRmOCUyMCU3RCUyMCUzRCU1Q24lMjAlMjB1bmVudkJ1ZmZlciUzQiU1Q24lNUNuZXhwb3J0JTIwZGVmYXVsdCUyMCU3QiU1Q24lMjAlMjAuLi51bmVudkJ1ZmZlciUyQyU1Q24lMjAlMjBCdWZmZXIlMkMlMjBTbG93QnVmZmVyJTVDbiU3RCUzQiU1Q24lNUNuJTIyJTJDJTIyaXNFbnRyeSUyMiUzQWZhbHNlJTJDJTIybmFtZSUyMiUzQSUyMmJ1ZmZlci1jbG91ZGZsYXJlLXBvbHlmaWxsLmpzJTIyJTdEJTJDJTdCJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGNsYXNzJTIwQnVmZmVyJTIwJTdCJTVDbiUyMCUyMG5vZGUlMjAlM0QlMjAndW5lbnYlMjBCdWZmZXInJTNCJTVDbiU3RCU1Q25leHBvcnQlMjBjb25zdCUyMGlzQXNjaWklMjAlM0QlMjAoKSUyMCUzRCUzRSUyMCU3QiUyMHJldHVybiUyMCdpc0FzY2lpJyUzQiUyMCU3RCUzQiU1Q25leHBvcnQlMjBjb25zdCUyMGlzVXRmOCUyMCUzRCUyMCgpJTIwJTNEJTNFJTIwJTdCJTIwcmV0dXJuJTIwJ2lzVXRmOCclM0IlMjAlN0QlM0IlNUNuJTVDbmV4cG9ydCUyMGRlZmF1bHQlMjAlN0IlNUNuJTIwJTIwQnVmZmVyJTJDJTVDbiUyMCUyMGlzQXNjaWklMkMlNUNuJTIwJTIwaXNVdGY4JTVDbiU3RCUzQiUyMiUyQyUyMmlzRW50cnklMjIlM0FmYWxzZSUyQyUyMm5hbWUlMjIlM0ElMjJ1bmVudi1ub2RlLWJ1ZmZlci1wb2x5ZmlsbC5qcyUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJvdXRwdXQlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiU3RCUyQyUyMnRyZWVzaGFrZSUyMiUzQXRydWUlN0QlN0Q=)
- [fixed](https://rollupjs.org/repl/?version=4.18.0&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMiUyRiUyRiUyMFRSRUUtU0hBS0lORyUyMC0lMjBkZXN0cnVjdHVyZWQlMjBpbXBvcnRzJTIwLSUyMGZpeGVkJTVDbmltcG9ydCUyMCU3QiUyMEJ1ZmZlciUyQyUyMGlzQXNjaWklMjAlN0QlMjBmcm9tJTIwJy4lMkZidWZmZXItY2xvdWRmbGFyZS1wb2x5ZmlsbC5qcyclM0IlNUNuJTVDbiUyRiUyRiUyMHNob3VsZCUyMHJldGFpbiUyMG9ubHklMjAlNjBpc0FzY2lpJTYwJTIwZnJvbSUyMHVuZW52LW5vZGUtYnVmZmVyLXBvbHlmaWxsLmpzJTVDbiUyRiUyRiUyMHRoZXJlJTIwaXMlMjBzb21lJTIwZXh0cmElMjBmbHVmZiUyMGZyb20lMjBidWZmZXItY2xvdWRmbGFyZS1wb2x5ZmlsbC5qcyUyMGJ1dCUyMGl0J3MlMjBtaW5pbWFsJTVDbmNvbnNvbGUubG9nKEJ1ZmZlciUyQyUyMGlzQXNjaWkpJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlMkMlMjJuYW1lJTIyJTNBJTIybWFpbi5qcyUyMiU3RCUyQyU3QiUyMmNvZGUlMjIlM0ElMjJpbXBvcnQlMjB1bmVudkJ1ZmZlciUyMGZyb20lMjAnLiUyRnVuZW52LW5vZGUtYnVmZmVyLXBvbHlmaWxsLmpzJyUzQiU1Q24lNUNuY29uc3QlMjB3b3JrZXJkQnVmZmVyJTIwJTNEJTIwcHJvY2Vzcy5nZXRCdWlsdGluTW9kdWxlKCU1QyUyMm5vZGUlM0FidWZmZXIlNUMlMjIpJTNCJTVDbiU1Q25leHBvcnQlMjBjb25zdCUyMCU3QiUyMEJ1ZmZlciUyQyUyMFNsb3dCdWZmZXIlN0QlMjAlM0QlNUNuJTIwJTIwd29ya2VyZEJ1ZmZlciUzQiU1Q24lNUNuZXhwb3J0JTIwJTdCJTIwaXNBc2NpaSUyQyUyMGlzVXRmOCUyMCU3RCU1Q24lMjAlMjBmcm9tJTIwJy4lMkZ1bmVudi1ub2RlLWJ1ZmZlci1wb2x5ZmlsbC5qcyclM0IlNUNuJTVDbmV4cG9ydCUyMGRlZmF1bHQlMjAlN0IlNUNuJTIwJTIwLi4udW5lbnZCdWZmZXIlMkMlNUNuJTIwJTIwQnVmZmVyJTJDJTIwU2xvd0J1ZmZlciU1Q24lN0QlM0IlNUNuJTVDbiUyMiUyQyUyMmlzRW50cnklMjIlM0FmYWxzZSUyQyUyMm5hbWUlMjIlM0ElMjJidWZmZXItY2xvdWRmbGFyZS1wb2x5ZmlsbC5qcyUyMiU3RCUyQyU3QiUyMmNvZGUlMjIlM0ElMjJleHBvcnQlMjBjbGFzcyUyMEJ1ZmZlciUyMCU3QiU1Q24lMjAlMjBub2RlJTIwJTNEJTIwJ3VuZW52JTIwQnVmZmVyJyUzQiU1Q24lN0QlNUNuZXhwb3J0JTIwY29uc3QlMjBpc0FzY2lpJTIwJTNEJTIwKCklMjAlM0QlM0UlMjAlN0IlMjByZXR1cm4lMjAnaXNBc2NpaSclM0IlMjAlN0QlM0IlNUNuZXhwb3J0JTIwY29uc3QlMjBpc1V0ZjglMjAlM0QlMjAoKSUyMCUzRCUzRSUyMCU3QiUyMHJldHVybiUyMCdpc1V0ZjgnJTNCJTIwJTdEJTNCJTVDbiU1Q25leHBvcnQlMjBkZWZhdWx0JTIwJTdCJTVDbiUyMCUyMEJ1ZmZlciUyQyU1Q24lMjAlMjBpc0FzY2lpJTJDJTVDbiUyMCUyMGlzVXRmOCU1Q24lN0QlM0IlMjIlMkMlMjJpc0VudHJ5JTIyJTNBZmFsc2UlMkMlMjJuYW1lJTIyJTNBJTIydW5lbnYtbm9kZS1idWZmZXItcG9seWZpbGwuanMlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyb3V0cHV0JTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlN0QlMkMlMjJ0cmVlc2hha2UlMjIlM0F0cnVlJTdEJTdE)

Namespaced import usage:
- [current](https://rollupjs.org/repl/?version=4.18.0&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMiUyRiUyRiUyMFRSRUUtU0hBS0lORyUyMC0lMjBuYW1lc3BhY2VkJTIwaW1wb3J0cyUyMC0lMjBicm9rZW4lNUNuaW1wb3J0JTIwYiUyMGZyb20lMjAnLiUyRmJ1ZmZlci1jbG91ZGZsYXJlLXBvbHlmaWxsLmpzJyUzQiU1Q24lNUNuJTJGJTJGJTIwc2hvdWxkJTIwcmV0YWluJTIwb25seSUyMCU2MGlzQXNjaWklNjAlMjBmcm9tJTIwdW5lbnYtbm9kZS1idWZmZXItcG9seWZpbGwuanMlNUNuJTJGJTJGJTIwYnV0JTIwcmV0YWlucyUyMGFsbCUyMGR1ZSUyMHRvJTIwJTYwaW1wb3J0JTIwdW5lbnZCdWZmZXIlMjBmcm9tJTIwJy4lMkZ1bmVudi1ub2RlLWJ1ZmZlci1wb2x5ZmlsbC5qcyclM0IlNjAlNUNuY29uc29sZS5sb2coYi5CdWZmZXIlMkMlMjBiLmlzQXNjaWkpJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlMkMlMjJuYW1lJTIyJTNBJTIybWFpbi5qcyUyMiU3RCUyQyU3QiUyMmNvZGUlMjIlM0ElMjJpbXBvcnQlMjB1bmVudkJ1ZmZlciUyMGZyb20lMjAnLiUyRnVuZW52LW5vZGUtYnVmZmVyLXBvbHlmaWxsLmpzJyUzQiU1Q24lNUNuY29uc3QlMjB3b3JrZXJkQnVmZmVyJTIwJTNEJTIwcHJvY2Vzcy5nZXRCdWlsdGluTW9kdWxlKCU1QyUyMm5vZGUlM0FidWZmZXIlNUMlMjIpJTNCJTVDbiU1Q25leHBvcnQlMjBjb25zdCUyMCU3QiUyMEJ1ZmZlciUyQyUyMFNsb3dCdWZmZXIlN0QlMjAlM0QlNUNuJTIwJTIwd29ya2VyZEJ1ZmZlciUzQiU1Q24lNUNuZXhwb3J0JTIwJTdCJTIwaXNBc2NpaSUyQyUyMGlzVXRmOCUyMCU3RCU1Q24lMjAlMjBmcm9tJTIwJy4lMkZ1bmVudi1ub2RlLWJ1ZmZlci1wb2x5ZmlsbC5qcyclM0IlNUNuJTVDbmV4cG9ydCUyMGRlZmF1bHQlMjAlN0IlNUNuJTIwJTIwLi4udW5lbnZCdWZmZXIlMkMlNUNuJTIwJTIwQnVmZmVyJTJDJTIwU2xvd0J1ZmZlciU1Q24lN0QlM0IlNUNuJTVDbiUyMiUyQyUyMmlzRW50cnklMjIlM0FmYWxzZSUyQyUyMm5hbWUlMjIlM0ElMjJidWZmZXItY2xvdWRmbGFyZS1wb2x5ZmlsbC5qcyUyMiU3RCUyQyU3QiUyMmNvZGUlMjIlM0ElMjJleHBvcnQlMjBjbGFzcyUyMEJ1ZmZlciUyMCU3QiU1Q24lMjAlMjBub2RlJTIwJTNEJTIwJ3VuZW52JTIwQnVmZmVyJyUzQiU1Q24lN0QlNUNuZXhwb3J0JTIwY29uc3QlMjBpc0FzY2lpJTIwJTNEJTIwKCklMjAlM0QlM0UlMjAlN0IlMjByZXR1cm4lMjAnaXNBc2NpaSclM0IlMjAlN0QlM0IlNUNuZXhwb3J0JTIwY29uc3QlMjBpc1V0ZjglMjAlM0QlMjAoKSUyMCUzRCUzRSUyMCU3QiUyMHJldHVybiUyMCdpc1V0ZjgnJTNCJTIwJTdEJTNCJTVDbiU1Q25leHBvcnQlMjBkZWZhdWx0JTIwJTdCJTVDbiUyMCUyMEJ1ZmZlciUyQyU1Q24lMjAlMjBpc0FzY2lpJTJDJTVDbiUyMCUyMGlzVXRmOCU1Q24lN0QlM0IlMjIlMkMlMjJpc0VudHJ5JTIyJTNBZmFsc2UlMkMlMjJuYW1lJTIyJTNBJTIydW5lbnYtbm9kZS1idWZmZXItcG9seWZpbGwuanMlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyb3V0cHV0JTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlN0QlMkMlMjJ0cmVlc2hha2UlMjIlM0F0cnVlJTdEJTdE)
- [fixed for real](https://rollupjs.org/repl/?version=4.18.0&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMiUyRiUyRiUyMFRSRUUtU0hBS0lORyUyMC0lMjBuYW1lc3BhY2VkJTIwaW1wb3J0cyUyMC0lMjBicm9rZW4lNUNuaW1wb3J0JTIwYiUyMGZyb20lMjAnLiUyRmJ1ZmZlci1jbG91ZGZsYXJlLXBvbHlmaWxsLmpzJyUzQiU1Q24lNUNuJTJGJTJGJTIwc2hvdWxkJTIwcmV0YWluJTIwb25seSUyMCU2MGlzQXNjaWklNjAlMjBmcm9tJTIwdW5lbnYtbm9kZS1idWZmZXItcG9seWZpbGwuanMlNUNuJTJGJTJGJTIwdGhlcmUlMjBpcyUyMHNvbWUlMjBleHRyYSUyMGZsdWZmJTIwZnJvbSUyMGJ1ZmZlci1jbG91ZGZsYXJlLXBvbHlmaWxsLmpzJTIwYnV0JTIwaXQncyUyMG1pbmltYWwlNUNuJTJGJTJGJTIwYWxzbyUyMCU2MGlzVXRmOCU2MCUyMGlzJTIwcmV0YWluZWQlMjAlNUNuY29uc29sZS5sb2coYi5CdWZmZXIlMkMlMjBiLmlzQXNjaWkpJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlMkMlMjJuYW1lJTIyJTNBJTIybWFpbi5qcyUyMiU3RCUyQyU3QiUyMmNvZGUlMjIlM0ElMjJjb25zdCUyMHdvcmtlcmRCdWZmZXIlMjAlM0QlMjBwcm9jZXNzLmdldEJ1aWx0aW5Nb2R1bGUoJTVDJTIybm9kZSUzQWJ1ZmZlciU1QyUyMiklM0IlNUNuJTVDbmV4cG9ydCUyMGNvbnN0JTIwJTdCJTIwQnVmZmVyJTJDJTIwU2xvd0J1ZmZlciU3RCUyMCUzRCU1Q24lMjAlMjB3b3JrZXJkQnVmZmVyJTNCJTVDbiU1Q25leHBvcnQlMjAlN0IlMjBpc0FzY2lpJTJDJTIwaXNVdGY4JTIwJTdEJTVDbiUyMCUyMGZyb20lMjAnLiUyRnVuZW52LW5vZGUtYnVmZmVyLXBvbHlmaWxsLmpzJyUzQiU1Q24lNUNuaW1wb3J0JTIwJTdCJTIwaXNBc2NpaSUyQyUyMGlzVXRmOCUyMCU3RCU1Q24lMjAlMjBmcm9tJTIwJy4lMkZ1bmVudi1ub2RlLWJ1ZmZlci1wb2x5ZmlsbC5qcyclM0IlNUNuJTVDbmV4cG9ydCUyMGRlZmF1bHQlMjAlN0IlNUNuJTIwJTIwaXNBc2NpaSUyQyUyMGlzVXRmOCUyQyU1Q24lMjAlMjBCdWZmZXIlMkMlMjBTbG93QnVmZmVyJTVDbiU3RCUzQiU1Q24lNUNuJTIyJTJDJTIyaXNFbnRyeSUyMiUzQWZhbHNlJTJDJTIybmFtZSUyMiUzQSUyMmJ1ZmZlci1jbG91ZGZsYXJlLXBvbHlmaWxsLmpzJTIyJTdEJTJDJTdCJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGNsYXNzJTIwQnVmZmVyJTIwJTdCJTVDbiUyMCUyMG5vZGUlMjAlM0QlMjAndW5lbnYlMjBCdWZmZXInJTNCJTVDbiU3RCU1Q25leHBvcnQlMjBjb25zdCUyMGlzQXNjaWklMjAlM0QlMjAoKSUyMCUzRCUzRSUyMCU3QiUyMHJldHVybiUyMCdpc0FzY2lpJyUzQiUyMCU3RCUzQiU1Q25leHBvcnQlMjBjb25zdCUyMGlzVXRmOCUyMCUzRCUyMCgpJTIwJTNEJTNFJTIwJTdCJTIwcmV0dXJuJTIwJ2lzVXRmOCclM0IlMjAlN0QlM0IlNUNuJTVDbmV4cG9ydCUyMGRlZmF1bHQlMjAlN0IlNUNuJTIwJTIwQnVmZmVyJTJDJTVDbiUyMCUyMGlzQXNjaWklMkMlNUNuJTIwJTIwaXNVdGY4JTVDbiU3RCUzQiUyMiUyQyUyMmlzRW50cnklMjIlM0FmYWxzZSUyQyUyMm5hbWUlMjIlM0ElMjJ1bmVudi1ub2RlLWJ1ZmZlci1wb2x5ZmlsbC5qcyUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJvdXRwdXQlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiU3RCUyQyUyMnRyZWVzaGFrZSUyMiUzQXRydWUlN0QlN0Q=)


So the last pattern is the way to go even though it's the most verbose. Oh well...

Reply:
This seems like the kind of optimization that rollup could handle automatically. Maybe one day. I got too spoiled by the capabilities of Closure Compiler at Google.